includes/class-sensei-course-enrolment-manager.php

```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * File containing the class Sensei_Course_Enrolment_Manager.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Handles the management of course enrolment.
+ */
+class Sensei_Course_Enrolment_Manager {
+	const COURSE_ENROLMENT_SITE_SALT_OPTION = 'sensei_course_enrolment_site_salt';
+
+	/**
+	 * All course enrolment providers.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private static $enrolment_providers;
+
+	/**
+	 * Gets the descriptive name of the provider by ID.
+	 *
+	 * @param string $provider_id Unique identifier of the enrolment provider.
+	 *
+	 * @return string|false
+	 */
+	public static function get_enrolment_provider_name_by_id( $provider_id ) {
+		$provider = self::get_enrolment_provider_by_id( $provider_id );
+		if ( ! $provider ) {
+			return false;
+		}
+
+		$provider_class = get_class( $provider );
+
+		return $provider_class::get_name();
+	}
+
+	/**
+	 * Gets the enrolment provider object by its ID.
+	 *
+	 * @param string $provider_id Unique identifier of the enrolment provider.
+	 *
+	 * @return Sensei_Course_Enrolment_Provider_Interface|false
+	 */
+	public static function get_enrolment_provider_by_id( $provider_id ) {
+		$all_providers = self::get_all_enrolment_providers();
+		if ( ! isset( $all_providers[ $provider_id ] ) ) {
+			return false;
+		}
+
+		return $all_providers[ $provider_id ];
+	}
+
+	/**
+	 * Check if we should use the legacy enrolment check. Legacy check uses course
+	 * progress to determine enrolment.
+	 *
+	 * @return bool
+	 */
+	public static function use_legacy_enrolment_check() {
+		$use_legacy_enrolment_check = false;
+
+		// Check if WCPC is around but not offering enrolment providers (an old version).
+		if (
+			class_exists( '\Sensei_WC_Paid_Courses\Sensei_WC_Paid_Courses' ) &&
+			! class_exists( '\Sensei_WC_Paid_Courses\Course_Enrolment_Providers' )
+		) {
+			$use_legacy_enrolment_check = true;
+		}
+
+		return apply_filters( 'sensei_legacy_enrolment_check', $use_legacy_enrolment_check );
+	}
+
+	/**
+	 * Get an array of all the instantiated course enrolment providers.
+	 *
+	 * @return Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	public static function get_all_enrolment_providers() {
+		if ( ! isset( self::$enrolment_providers ) ) {
+			self::$enrolment_providers = [];
+
+			/**
+			 * Fetch all registered course enrolment providers.
+			 *
+			 * @param string[] $provider_classes List of enrolment providers classes.
```

Comment:
I think this param is an array of classes that implements the Sensei_Course_Enrolment_Provider_Interface, is it not?

Reply:
It is, but I _think_ `Sensei_Course_Enrolment_Provider_Interface[]` would be a typehint for an array of instantiated objects implementing `Sensei_Course_Enrolment_Provider_Interface`.

Reply:
Hum... I got! So, in this case, would it not be better to use something like: `@param callable[]`? 🤔
I think because reading it we can think it's the string of the class name.

Reply:
I don't think we can hint any more that it is an array of class names (besides the description). These are just arrays of strings, not `callable` types ([ref](https://www.php.net/manual/en/language.types.callable.php)).

Reply:
I got! Thank you @jom ! =)